Fluentd supports logging to file

mburke5678 opened the pull request:
https://jira.coreos.com/browse/OSDOCS-8

[Fluentd supports logging to file](https://github.com/openshift/origin-aggregated-logging/pull/1336)

mburke5678 commented:
@VladMasarik PTAL

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +937,43 @@ existing `journald` log driver configuration. It is recommended to use the
+*Configuring Fluentd to Send Logs to a File*
+
+Fluentd writes logs into a file specified by the `LOGGING_FILE_PATH` environment variable, by default `/var/log/fluentd/fluentd.log`. 
+Logs are stored with the older log listed first. You can configure Fluentd to output logs to the console (`STDOUT`), 
```
@VladMasarik I got this from your MD file:
"Contents of files stored in the directory where `LOGGING_FILE_PATH` is, are printed out. Starting with the oldest log. "
Did I interpret this incorrectly?

VladMasarik replied:
Oh, this one. By that I meant, that `fluentd.log` contains newest logs but when using `logs` utility you start printing oldest log and continue to newer one. So when you type `logs` into a terminal the output will be "sorted" starting from the oldest log on top and newest log on the bottom. And I have no idea how to express this behavior.

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +937,43 @@ existing `journald` log driver configuration. It is recommended to use the
+
+| `LOGGING_FILE_PATH` | The path to the Fluentd log file. If you want Fluentd to output to the console, `STDOUT`, set this variable to `console` value. The default value is `/var/log/fluentd/fluentd.log`.
+| `LOGGING_FILE_AGE` | The length of time in days the Fluentd retains logs before deleting. The default value is `10`.
+| `LOGGING_FILE_SIZE` | The maximum size of a Fluentd log file in bytes. If the size of the log file is exceeds this value, the log file gets rotated. The default is 1MB.
```
@VladMasarik Apparently, your English is better than mine, and I am a native speaker!

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +937,50 @@ existing `journald` log driver configuration. It is recommended to use the
+You can configure Fluentd to output logs to the console (`STDOUT`), by setting the `LOGGING_FILE_PATH` variable to `console` value.
+
+To set the maximum size of a single log file, use the `LOGGING_FILE_SIZE` parameter or environment variable. To specify how many log files are retained 
+before {product-title} deletes the files, use the `LOGGING_FILE_AGE` parameter or environment variables.
```
> {product-title}

Does this get replaced later?

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +937,50 @@ existing `journald` log driver configuration. It is recommended to use the
+For example:
+
+----
+oc exec $fluentd-pod -- logs
```
> --

Should those two dashes be there?

mburke5678 commented:
@VladMasarik Forgive my ignorance. In the current OpenShift environment, fluentd collects the log data and users can view the data in ES. With the log utility, the user can also view the logs directly in the console? 
When the logs are rotated, what happens to the older logs? Are they kept and compressed (for viewing with logs), or removed?
Thank you for all your help!

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +949,63 @@ existing `journald` log driver configuration. It is recommended to use the
+
+To control the size of the logs, you can specify a maximum log size and the length of time that {product-title} retains the logs. 
+ 
+* The `LOGGING_FILE_SIZE` parameter specifies how large the file can be before {product-title} rotates the files. When the log reaches the specified size, {product-title} automatically compresses and renames the *fluentd.log* log file so that new logging data can be collected.  
```
> {product-title} automatically compresses and renames

There is no compression, OpenShift just renames the files. It is possible to compress them but we don't do it.

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +949,63 @@ existing `journald` log driver configuration. It is recommended to use the
+
+* The `LOGGING_FILE_AGE` parameter specifies how many of the compressed log files can be retained before {product-title} deletes the files. 
+
+The following example shows logs in a cluster where `LOGGING_FILE_SIZE=1Mb` and `LOGGING_FILE_AGE=4`. When the *fluentd.log* reaches 1Mb, {product-title} deletes the current *fluentd.log.4*, renames the *fluentd.log* and *fluentd.log.1* and renames the other files in turn.  
```
> `LOGGING_FILE_SIZE=1Mb`

Should be `LOGGING_FILE_SIZE=1024000`

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +949,63 @@ existing `journald` log driver configuration. It is recommended to use the
+
+| `LOGGING_FILE_PATH` | Speciy the path to the Fluentd log file. If you want Fluentd to output to the console, `STDOUT`, set this variable to `console`. The default value is `/var/log/fluentd/fluentd.log`.
+| `LOGGING_FILE_AGE` | The number of logs that Fluentd retains before deleting. The default value is `10`.
+| `LOGGING_FILE_SIZE` | The number of log files that are retained before {product-title} deletes the files. If the number of the log file exceeds this value, the log file gets rotated. The default is 1MB.
```
> The number of log files that are retained before {product-title} deletes the files. If the number of the log file exceeds this value, the log file gets rotated. The default is 1MB.

Can you replace this text by the following?:

Is a maximum size of a single log file in Bytes. If this value is exceeded by `fluentd.log`, log files are renamed accordingly and new `fluentd.log` is created. If there are more log files than is a value of `LOGGING_FILE_AGE` the oldest log gets deleted (Log rotation). Default value is `1024000` which means it is 1 MB.

VladMasarik replied:
Well, if it makes sense to an English speaker.

VladMasarik commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +949,63 @@ existing `journald` log driver configuration. It is recommended to use the
+----
+LOGGING_FILE_PATH=console
+LOGGING_FILE_AGE=30
+LOGGING_FILE_PATH=1024000
```
> LOGGING_FILE_PATH=1024000

I believe you wanted to say `SIZE` not `PATH`

mburke5678 replied:
Sorry. I was being sloppy here.

mburke5678 commented:
@QiaolingTang Please take a look. 
Changes related to the associated Jira start at [line 965](https://github.com/openshift/openshift-docs/pull/12586/files#diff-0bd2e2c1a5da6f157682d4e95ecea473R965).
The other changes are issues I noticed when working with the EFK stack.

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +962,65 @@ existing `journald` log driver configuration. It is recommended to use the
+----
+oc exec logging-es-data-master-2-deploy logs
+oc exec logging-kibana-1 logs
+----
```
These changes are in fluentd pod, not the kibana or elasticsearch, so the command should be "oc exec $fluentd-pod -- logs"

mburke5678 replied:
@VladMasarik @QiaolingTang Do we need the dashes before `logs`? There has been some conflicting information.

QiaolingTang replied:
@mburke5678 It doesn't matter. I tried "oc exec logging-fluentd-lmvms logs" and "oc exec logging-fluentd-lmvms -- logs", both of them are valid

VladMasarik replied:
I have just never seen anyone use the `--`. And since I had no way to test it at that time, I just didn't want to risk showing falsie information.

mburke5678 replied:
The command in the docs has no `--`. 
`oc exec logging-fluentd-lmvms logs`

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +962,65 @@ existing `journald` log driver configuration. It is recommended to use the
+LOGGING_FILE_PATH=console
+LOGGING_FILE_AGE=30
+LOGGING_FILE_SIZE=1024000
+----
```
I think it's better to illustrate how to set these parameters in the doc.

richm replied:
It doesn't make any sense to use `LOGGING_FILE_PATH=console` with `LOGGING_FILE_AGE` and `LOGGING_FILE_SIZE`

jcantrill replied:
I think it wold be better to simply exclude 'LOGGING_FILE_PATH' all together.  I would expect customer's to never really modify this location

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -937,6 +962,65 @@ existing `journald` log driver configuration. It is recommended to use the
+[[fluentd-file]]
+*Configuring Fluentd to Send Logs to a File*
+
+Fluentd writes logs to a file you specify with the `LOGGING_FILE_PATH` environment variable, by default `/var/log/fluentd/fluentd.log`.  
```
Fluentd writes its logs to console if LOGGING_FILE_PATH=console, it's not a file, so I think "Fluentd writes logs to a file you specify with the `LOGGING_FILE_PATH` environment variable" is not accurate.

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -622,6 +622,31 @@ process with:
+----
+
+<1> The Curator pod. Only one pod is needed for Curator.
+<2> The Elasticsearch pod on this host. 
```
This is an es deploy pod, not es pod. The curator pod and the es pod in this example are not deployed successfully.  If you want to show the whole EFK stack here, it's better to show a healthy EFK stack, or it may cause misunderstanding.

mburke5678 replied:
@QiaolingTang @VladMasarik  Do you have sample output that I can use.

VladMasarik replied:
I believe this should be OK:
```
$ oc get pod
NAME                                      READY     STATUS    RESTARTS   AGE
logging-curator-1-gqbwl                   1/1       Running   0          2d
logging-es-data-master-soss6cj7-1-mvcfz   2/2       Running   0          2d
logging-fluentd-jwkb2                     1/1       Running   0          2d
logging-kibana-1-q5nkg                    2/2       Running   0          2d
```
although I am not sure if the number of kibana and ES pods is OK, but I think it is.

mburke5678 replied:
@VladMasarik Thank you. I tried to stand up the EFK stack to get the proper output, but was unable to because of memory limitations on my servers.

jcantrill replied:
I would add to this as well that in 3.11, curator was moved to a BatchJob so you will only see a curator pod when the job is executing.  Otherwise, you will only see the others.

mburke5678 commented:
@QiaolingTang Excellent comments. Thank you very much for your help.

mburke5678 commented:
@VladMasarik @QiaolingTang 
I reorganized the section into 3 sections. The section was not clear with all the information together:
_Viewing Fluentd Logs_
_Configuring Fluentd Log Location_
_Configuring Fluentd Log Rotation_

Final edits for above comments prior to the reorg are: de675ffeea19a25cb25f7d4592ea913f73265d79

mburke5678 commented:
@VladMasarik BUMP ^^

mburke5678 commented:
@VladMasarik @QiaolingTang 
I have updated the oc get pods -o wide output [as Vlad requested](https://github.com/openshift/openshift-docs/pull/12586#commitcomment-31405066).

I changed the way that users adjust log rotation to environment variables [as Qiaoling suggested](https://github.com/openshift/openshift-docs/pull/12586#commitcomment-31232840).

If @richm can confirm the steps to turn off log rotation, can we move this PR forward to peer review?

richm commented on install_config/aggregate_logging.adoc:
```diff
@@ -1032,17 +1033,10 @@ the xref:../install/configuring_inventory_file.html#configuring-ansible[default
-$ ansible-playbook [-i </path/to/inventory>] \
-    playbooks/openshift-logging/config.yml
-----
+To turn off log rotation, if the `LOGGING_FILE_PATH=console` is set and the `DEBUG` environment variable exists. With log rotation off, all logs will be printed to console but then they would be redirected to *_/var/log/fluentd.log file_*.
```
@jcantrill please confirm

jcantrill replied:
Wouldn't it be easier to say:
"Turn off log rotation by setting `LOGGING_FILE_PATH=console`.  This will cause fluentd to write logs to STDOUT where they can be retrieved using the normal 'oc logs $fpod' command"

I don't believe its important for anyone to know the logs will go anyplace else but STDOUT.  The presumption here, IMO, is that when you write to console, use `oc logs`, when writing to file, use `oc exec $fpod logs`

mburke5678 commented:
@jcantrill @richm @VladMasarik 
Is this PR related and to and waiting on [Fluentd pod should log its logs to /var/log/fluentd/fluentd.log by default ](https://bugzilla.redhat.com/show_bug.cgi?id=1649997)? 
That BZ appears to be 3.11 only. The [PR I have associated with this PR](https://github.com/openshift/origin-aggregated-logging/pull/1336) is backported to 3.9. I am not sure on what is needed here.

mburke5678 commented:
@jcantrill I believe I was waiting on your answer to [Rich's question](https://github.com/openshift/openshift-docs/pull/12586#discussion_r237239371), but got caught up in the 4.0 docs conversion project.

mburke5678 commented:
@openshift/team-documentation PTAL

mburke5678 commented:
/cherrypick enterprise-3.11

mburke5678 commented:
/cherrypick enterprise-3.10

mburke5678 commented:
/cherrypick enterprise-3.9
